File: LP2CPP/Nqueens.cpp
#include <iostream>
#include <vector>

// Forward declaration of print_board function
void print_board(const std::vector<std::vector<int>>& board, int n);

bool is_safe(const std::vector<std::vector<int>>& board, int row, int col, int n) {
    // Check same column
    for (int i = 0; i < row; ++i) {
        if (board[i][col] == 1) {
            return false;
        }
    }
    
    // Check upper-left diagonal
    for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; --i, --j) {
        if (board[i][j] == 1) {
            return false;
        }
    }
    
    // Check upper-right diagonal
    for (int i = row - 1, j = col + 1; i >= 0 && j < n; --i, ++j) {
        if (board[i][j] == 1) {
            return false;
        }
    }
    
    return true;
}

bool solve_n_queens(std::vector<std::vector<int>>& board, int row, int n) {
    if (row == n) {
        print_board(board, n);
        return true;
    }
    
    for (int col = 0; col < n; ++col) {
        if (is_safe(board, row, col, n)) {
            board[row][col] = 1; // Place Queen
            if (solve_n_queens(board, row + 1, n)) {
                return true;
            }
            board[row][col] = 0; // Backtrack
        }
    }
    
    return false;
}

void print_board(const std::vector<std::vector<int>>& board, int n) {
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            std::cout << (board[i][j] ? 'Q' : '.') << ' ';
        }
        std::cout << std::endl;
    }
    std::cout << std::endl;
}

int main() {
    int n = 8;
    std::vector<std::vector<int>> board(n, std::vector<int>(n, 0));
    
    if (!solve_n_queens(board, 0, n)) {
        std::cout << "No solution exists." << std::endl;
    }
    
    return 0;
}

File: LP2/nQueens.py
def is_safe(board, row, col, n):
    # Check same column
    for i in range(row):
        if board[i][col] == 1:
            return False

    # Check upper-left diagonal
    for i, j in zip(range(row-1, -1, -1), range(col-1, -1, -1)):
        if board[i][j] == 1:
            return False

    # Check upper-right diagonal
    for i, j in zip(range(row-1, -1, -1), range(col+1, n)):
        if board[i][j] == 1:
            return False

    return True

def solve_n_queens(board, row, n):
    if row == n:
        print_board(board, n)
        return True

    for col in range(n):
        if is_safe(board, row, col, n):
            board[row][col] = 1  # Place Queen
            if solve_n_queens(board, row + 1, n):
                return True
            board[row][col] = 0  # Backtrack

    return False

def print_board(board, n):
    for i in range(n):
        for j in range(n):
            print('Q' if board[i][j] else '.', end=' ')
        print()
    print()

# Example usage
n = 8
board = [[0 for _ in range(n)] for _ in range(n)]
if not solve_n_queens(board, 0, n):
    print("No solution exists.")
